### PR TITLE
Requestify Inferring Generic Requirements

### PR DESCRIPTION
**Pull request by CodaFi**
Unify a bunch of quasi-independent callsites into a single request that
builds up a generic signature from a variety of inference sources.  This
draws the extension typealias workaround for SE-0229 into TypeCheckDecl
where we can better work on refactoring it.  The goal is to use this as
a springboard to requests that compute generic environments for various
and sundry decls.

### CODE REVIEW

**Comment by CodaFi** [comment]
Unfortunately, only code structure improvements in this one.  No perf improvements from laziness - seems like we were already doing a pretty good job of minimizing our construction of GSBs here.

**Comment by CodaFi** [comment]
@swift-ci please test

File: lib/Sema/TypeCheckDecl.cpp [hunk 1]
```diff
@@ -4392,6 +4392,17 @@ static Type formExtensionInterfaceType(
...
 
+/// Retrieve the generic parameter depth of the extended type.
+static unsigned getExtendedTypeGenericDepth(ExtensionDecl *ext) {
+  auto nominal = ext->getSelfNominalTypeDecl();
+  if (!nominal) return static_cast<unsigned>(-1);
+
+  auto sig = nominal->getGenericSignatureOfContext();
+  if (!sig) return static_cast<unsigned>(-1);
+
+  return sig->getGenericParams().back()->getDepth();
```
Comment (Azoy) [inline lib/Sema/TypeCheckDecl.cpp]: `return nominal->getGenericContextDepth();` ?

Comment (CodaFi) [reply lib/Sema/TypeCheckDecl.cpp]: I don't think we need a context walk here if we already have this information available in this (admittedly baroque) form.  There's quite a few places in the compiler that are computing depths this way.  Perhaps we should look into unifying them.

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
@swift-ci please test

**Comment by CodaFi** [comment]
⛵️

**Pull request merged**